Native browser CompressionStream (gzip) for saves

Zarniwoops opened the pull request:
This uses the native CompressionStream (async pain in the ass) instead of doing compression in pure js.

It's 10 times faster, and surprisingly, a bit smaller.

I don't think I made too much of a mess with async, but it's not as clean as I'd like.

Zarniwoops committed:
Add option to compress save games with native browser CompressionStream (gzip). Slightly smaller, 10x faster

Zarniwoops committed:
Fixed load game preview to use new format. Added support for RFC-compliant data url save format

YukiMonsai commented:
ah, I totally missed this
Is this different from https://github.com/Ada18980/KinkiestDungeon/pull/223 ?

Zarniwoops commented:
The main difference with this one is that I was trying to keep the changes as minimal as I could. It's also opt-out by default. You have to enable it in the config options.

Ada18980 commented:
Ill make a few changes I think, but @ewhac I do think a minimalist approach is preferred here

ewhac commented on Game/src/base/KinkyDungeon.ts:
```diff
@@ -7204,6 +7244,84 @@ function KDDrawGameSetupTabs(_xOffset: number = 0) {
+		binary += String.fromCharCode(compressed[i]);
+	}
+	// return 'data:application/vnd.straightlaced.kinkydungeon.save.game+gzip;version=2;base64,' + btoa(binary);
+	return 'gzip:' + btoa(binary);
```
Couple of things:
* You're basically creating a second copy of the compressed data in the conversion from `Uint8Array` to a string.  This copy is also twice as large, as JS strings are in UTF16.  Recall that I have save codes as much as 7 MiB in size after gzip compression (I can share them with you for testing if you like).  I suggest using `Uint8Array.prototype.toBase64()`, to encode the compressed `Uint8Array` directly, as recommended by [Mozilla's docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Uint8Array/toBase64).
* ***Strongly*** recommend generating a data URL with a full MIME type.  This will allow both the user and the code to immediately distinguish between save code types (yes, you *will* be expanding to outfits/wardrobe in the future), as well as add a version field letting you evolve the save format over time.

ewhac reviewed (commented):
Comment.

Zarniwoops committed:
Use RFC 2397 data URL with vendor MIME type for gzip saves

Use Uint8Array.toBase64() instead of String.fromCharCode loop + btoa(),
avoiding a redundant UTF-16 copy of the compressed data.

Loader validates save type from MIME when present, rejects non-game
saves, and falls back to LZString for classic save codes.

Zarniwoops commented:
Alright, I updated it.

Zarniwoops committed:
Use toBase64() with btoa fallback for older engines

Zarniwoops commented:
toBase64() doesn't work on the current electron bundle of the game. So now it's auto-detecting and using the best method.

ewhac commented on out/saveworker.js:
```diff
@@ -1,8 +1,55 @@
+		offset += chunk.length;
+	}
+
+	// Use toBase64() where available (Chrome 137+), fall back to btoa for older engines
+	let b64;
+	if (typeof compressed.toBase64 === 'function') {
+		b64 = compressed.toBase64();
```
Nice.  But, as you point out, `toBase64()` is crazy-new (sorry; didn't know that when I suggested it).

However, `readAsDataURL()` has been around since Chrome 6 for [`FileReader`](https://developer.mozilla.org/en-US/docs/Web/API/FileReader/readAsDataURL), ~~and Chrome 7 for [`FileReaderSync`](https://developer.mozilla.org/en-US/docs/Web/API/FileReaderSync/readAsDataURL).  In fact, I think I'm going to re-work my own version to use `FileReaderSync`, since it'll get rid of at least one `await`.~~  Crap; `FireReaderSync()` is only available inside WebWorkers.

Lemme noodle on this for a bit...

ewhac commented:
Here's my implementation, which is basically a copy of what I wrote for `super_saver` (couldn't really find a way to simplify it):

```ts
/**
 * Compress a string using native browser gzip, returns a data: URI with MIME type
 */
async function KDCompressGzip (input: string): Promise<string> {
	const mime_type = 'application/vnd.straightlaced.kinkydungeon.save.game+gzip;version=2';

	const encoder = new TextEncoder();
	// b_js -- blob of JSON
	const save_b_js = new Blob ([encoder.encode (input)]);
	const save_pipe = save_b_js.stream().pipeThrough (new CompressionStream ('gzip'));

	const resp = new Response (save_pipe, { headers: [[ "Content-Type", mime_type ]]});
	// b_z -- blob compressed
	const save_b_z = await resp.blob();

	// Snarfed from MDN Web docs.  We do not catch the failure case.
	async function toBase64DataURL (blob: Blob): Promise<string> {
		return await new Promise ((resolve, reject) => {
			const reader = Object.assign (new FileReader(), {
				onload:  () => resolve (reader.result as string),
				onerror: () => reject (reader.error)
			});
			reader.readAsDataURL (blob);
		});
	}

	// Maybe return the Promise directly?
	return await toBase64DataURL (save_b_z);
}
```

Only significant difference from my version is that I catch errors from `toBase64DataURL()` and fall back to LZString.

All the APIs used are sufficiently old that anyone using an even older browser has far more problems going on.  That said, I have no wish to antagonize anyone by harping on this any further.
